### PR TITLE
Use API file in new location, allows overriding composite ops.

### PR DESCRIPTION
**Pull request by ailzhang**
Companion diff for https://github.com/pytorch/pytorch/pull/45241
There're two benefits coming after this change:
- if pytorch has a `math`(aka composite) formula defined for an op, xla will by default use it if it doesn't want to register an explicit lowering. 
  In the past we are required to lower some ops since we move an op from composite to device dispatch in pytorch core, after this PR it should be no longer required. (we just move the composite implementation to `math` formula in pytorch so that xla can continue using it).
- XLA now can optionally override composite kernels. No matter if an op is composite or not, if XLA register to it, it'll be used when dispatching to XLA.

### CODE REVIEW

**Review by JackCaoG — APPROVED** [review]
Thanks!

File: scripts/gen.py [hunk 1]
```diff
@@ -1033,27 +1034,37 @@ def generate_registrations(fgens, overrides):
...
-# For other non-leaf nodes, PyTorch covers both forward and backward for
-# them. But XLA can still optionally override them as necessary.
-def requires_registration(fgen):
-  return (fgen.leaf and not fgen.math
-         ) or fgen.mapsig in _FN_AUTOGRAD_XLA or fgen.func in _FN_AUTOGRAD_XLA
-
-
-def generate_functions(fgens):
+# For an op that requires_lowering=True:
+#   - If XLA has a lowering in AtenXlaType, we use register that kernel to XLA dispatch key.
```
Comment (davidel) [inline scripts/gen.py:1038]: `use register`? 😄

File: scripts/gen.py [hunk 2]
```diff
@@ -1033,27 +1034,37 @@ def generate_registrations(fgens, overrides):
...
+#   - If XLA doesn't have a lowering in AtenXlaType, we generate one in AtenXlaTypeDefault
+#     and register that kernel to XLA dispatch key.
+# For an op that requires_lowering=False, has_xla_lowering=True:
+#   - xla lowering in AtenXlaType will be registerd to XLA dispatch key and used.
+#     (note this is new  since it was impossible to override composite op before)
```
Comment (davidel) [inline scripts/gen.py:1043]: Drop one space after `new`

File: scripts/gen.py [hunk 3]
```diff
@@ -1033,27 +1034,37 @@ def generate_registrations(fgens, overrides):
...
+# For an op that has_autogradxla=True:
+#   - the kernel must have both forward and backward implemented and it's registered to
+#     AutogradXLA dispatch key.
+#     Backends should only use this when Autograd kernel in PyTorch codebase doesn't fit.
+#     E.g max_pool2d which enforce materializing indices for backward pass to use.
+def requires_registration(fgen, overrides):
+  requires_lowering = (fgen.dispatch and not fgen.math)
```
Comment (davidel) [inline scripts/gen.py:1050]: I am not sure we need parenthesis here.

Nit: The code below could be rewritten to allow `OR` shortcutting (starting from the more unlikely condition), but naming the single conditions clearly has its value.

**Review by davidel — APPROVED** [review]
Couple of minor nits.

File: scripts/gen.py (hunk 3 above)
Comment (ailzhang) [reply scripts/gen.py:1050]: Yep I rewrite the shortcut so that it's has names and easier to match back to the note :D

**Pull request merged**